Add `constraints` field to `TransactionLayer`

jonathanknowles opened the pull request:
# Issue Number

ADP-839

# Overview

This PR:

- [x] adds field `constraints` of type `TxConstraints` to the `TransactionLayer` record type.
- [x] adds an implementation of `constraints` to `Shelley.Transaction.TransactionLayer`.
- [x] adds property tests to demonstrate the correctness of `Shelley.Transaction.TransactionLayer.constraints`.

# Refactoring

This PR also:

- [x] makes `TxSize` into a concrete type exported by `Primitive.Types.Tx`.
- [x] removes `calcMinimumCoinValue` from `TransactionLayer`, as `TxConstraints.txOutputMinimumAdaQuantity` makes this redundant.

# Future work

This PR will make it possible to:

- remove `tokenBundleSizeAssessor` from `TransactionLayer`, as `TxConstraints.txOutputMaximumSize` makes this redundant.

jonathanknowles committed:
[Refactoring] Simplify representation of inputs in `SelectionSkeleton`.

The purpose of `SelectionSkeleton` is to provide all the data necessary
to calculate the fee associated with a selection.

However, this calculation only depends on the number of inputs, and not
the contents of each input.

Therefore, it's only necessary to include a count of the selected inputs
within `SelectionSkeleton`, and unnecessary to include the actual inputs
themselves.

jonathanknowles committed:
[Refactoring] Make fields of `SelectionSkeleton` share common `skeleton` prefix.

This makes the record field names a bit more consistent.

jonathanknowles committed:
[Refactoring] Make `TxSize` into a concrete type exported by `Primitive.Tx`.

Ultimately, transaction sizes are all expressible in terms of natural
numbers, regardless of whether the sizes are expressed in terms of
bytes, or in terms of some other unit.

So there's currently little need for a type class abstraction here, when
a newtype will suffice.

jonathanknowles committed:
Add `constraints` field to `TransactionLayer`.

The `constraints` field provides a value of `TxConstraints`.

jonathanknowles committed:
[Refactoring] Stop using `calcMinimumCoinValue` in `Wallet`.

Replace these uses with `constraints.txOutputMinimumAdaQuantity`.

jonathanknowles committed:
[Refactoring] Remove `calcMinimumCoinValue` from `TransactionLayer`.

This is made redundant by `constraints`.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -799,9 +797,8 @@ estimateTxSize witnessTag ctx (SelectionSkeleton inputCount outs chgs) =
         + sizeOf_Address address
-        + sizeOf_SmallArray
-        + sizeOf_Coin (TokenBundle.getCoin tokens)
-        + F.foldl' (\t -> (t +) . sizeOf_NativeAsset) 0 (TokenBundle.getAssets tokens)
```
Sounds good, but does it not risk being slower?

jonathanknowles replied:
> Sounds good, but does it not risk being slower?

I eventually decided to get rid of this. I'm pretty sure that our estimate is quite a bit higher than the real serialized length (for Shelley), but I'd be happier changing this function when we have a test to check that `estimateTxSize` gives answers that are greater than or equal to the results of serializing a transaction with `cardano-node` functions.

jonathanknowles committed:
[Refactoring] Introduce `TxSkeleton` type.

This record type includes the minimal set of fields that are absolutely
necessary for the `estimateTxSize` function to perform its calculations,
and nothing else.

As such, the fields of `TxSkeleton` contain a subset of the data included
in the union of `SelectionSkeleton` and `TransactionCtx`.

Using this type means that future tests for `estimateTxSize` will only
need to supply values that will actually be used by the function,
simplifying the effort to write tests.

jonathanknowles committed:
[Refactoring] Extract out function `estimateTxCost`.

This will allow it to be tested in conjunction with `txConstraints`.

jonathanknowles committed:
[Reformatting] Tidy function `estimateTxSize`.

jonathanknowles committed:
Add basic property tests for `txConstraints`.

jonathanknowles committed:
Add property test for `txOutputMaximumSize`.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -540,45 +550,82 @@ _decodeSignedTx era bytes = do
         _ ->
             Left ErrDecodeSignedTxNotSupported
 
+data TxSkeleton = TxSkeleton
```
Might as well put the
> his record type includes the minimal set of fields that are absolutely
necessary for the `estimateTxSize` function to perform its calculations,
and nothing else.
> 
> As such, the fields of `TxSkeleton` contain a subset of the data included
in the union of `SelectionSkeleton` and `TransactionCtx`.

From the commit message as a doc comment here.

jonathanknowles replied:
Ah good point! I'll add some comments in there. 👍🏻

jonathanknowles replied:
Fixed in d7ff662ccfd5c79531b047cbd37e7bc9f52080d0.

Anviking commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -704,3 +742,201 @@ dummyProtocolParameters = ProtocolParameters
+-- between 'txOutputSize' and 'txOutputMaximumSize' should also indicate that
+-- the bundle is oversized.
+--
+prop_txConstraints_txOutputMaximumSize :: Large TokenBundle -> Property
```
👍

Anviking commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -704,3 +742,201 @@ dummyProtocolParameters = ProtocolParameters
+-- selection produces a result that is consistent with the result of using
+-- 'estimateTxCost'.
+--
+prop_txConstraints_txCost :: MockSelection -> Property
```
👍

Anviking reviewed (approved):
lgtm!

jonathanknowles committed:
Add documentation comments for 'TxSkeleton' and related functions.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2620/files#r622191033

jonathanknowles committed:
Use `skeleton` for arguments of type `TxSkeleton`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2620/files#r622668819

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -546,45 +551,180 @@ _decodeSignedTx era bytes = do
+    }
+
+estimateTxCost :: ProtocolParameters -> TxSkeleton -> Coin
+estimateTxCost pp args =
```
```suggestion
estimateTxCost pp skeleton =
```

jonathanknowles replied:
Fixed in 59e10c52431f22d2998e5967b7a0595e7110f771.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -704,3 +742,201 @@ dummyProtocolParameters = ProtocolParameters
+-- the bundle is oversized.
+--
+prop_txConstraints_txOutputMaximumSize :: Large TokenBundle -> Property
+prop_txConstraints_txOutputMaximumSize (Large bundle) =
```
The `bundle` argument probably be blinded to avoid spamming the logs with the output of `show` for huge token bundles. (We already have the data we need in the counterexamples.)

jonathanknowles replied:
Fixed in 2099f94ebad72a8bdf1bd04b7e42120694075d32.

jonathanknowles committed:
Blind the argument to `prop_txConstraints_txOutputMaximumSize`.

We probably want to avoid spamming the test log with the output of
`show` for very large token bundles. We already have a large set of
counterexample data.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2620/files#r622669907
